src/crimson/os/seastore/segment_cleaner.h

```diff
@@ -162,6 +512,175 @@ class SegmentCleaner : public JournalSegmentProvider {
+    return config.segment_size * config.num_segments;
+  }
+
+  size_t get_unavailable_bytes() const {
+    return get_total_bytes() - get_available_bytes();
+  }
+
+  /// Returns bytes currently occupied by live extents (not journal)
+  size_t get_used_bytes() const {
+    return used_bytes;
+  }
+
+  /// Returns the number of bytes in unavailable segments that are not live
+  size_t get_reclaimable_bytes() const {
+    return get_unavailable_bytes() - get_used_bytes();
+  }
+
+  /**
+   * get_reclaim_ratio
+   *
+   * Returns the ratio of unavailable space that is not currently used.
+   */
+  double get_reclaim_ratio() const {
+    if (get_unavailable_bytes() == 0) return 0;
+    return (double)get_reclaimable_bytes() / (double)get_unavailable_bytes();
+  }
+
+  /**
+   * get_available_ratio
+   *
+   * Returns ratio of available space to write to total space
+   */
+  double get_available_ratio() const {
+    return (double)get_available_bytes() / (double)get_total_bytes();
+  }
+
+  /**
+   * get_immediate_bytes_to_gc_for_reclaim
+   *
+   * Returns the number of bytes to gc in order to bring the
+   * reclaim ratio below reclaim_ratio_usage_min.
+   */
+  size_t get_immediate_bytes_to_gc_for_reclaim() const {
+    if (get_reclaim_ratio() < config.reclaim_ratio_hard_limit)
+      return 0;
+
+    const size_t unavailable_target = std::max(
+      get_used_bytes() / (1.0 - config.reclaim_ratio_hard_limit),
+      (1 - config.reclaim_ratio_usage_min) * get_total_bytes());
+
+    if (unavailable_target > get_unavailable_bytes())
+      return 0;
+
+    return (get_unavailable_bytes() - unavailable_target) / get_reclaim_ratio();
+  }
+
+  /**
+   * get_immediate_bytes_to_gc_for_available
+   *
+   * Returns the number of bytes to gc in order to bring the
+   * the ratio of available disk space to total disk space above
+   * available_ratio_hard_limit.
+   */
+  size_t get_immediate_bytes_to_gc_for_available() const {
+    if (get_available_ratio() > config.available_ratio_hard_limit) {
+      return 0;
+    }
+
+    const double ratio_to_make_available = config.available_ratio_hard_limit -
+      get_available_ratio();
+    return ratio_to_make_available * (double)get_total_bytes()
+      / get_reclaim_ratio();
+  }
+
+  /**
+   * get_immediate_bytes_to_gc
+   *
+   * Returns number of bytes to gc in order to restore any strict
+   * limits.
+   */
+  size_t get_immediate_bytes_to_gc() const {
+    // number of bytes to gc in order to correct reclaim ratio
+    size_t for_reclaim = get_immediate_bytes_to_gc_for_reclaim();
+
+    // number of bytes to gc in order to correct available_ratio
+    size_t for_available = get_immediate_bytes_to_gc_for_available();
+
+    return std::max(for_reclaim, for_available);
+  }
+
+  void mark_closed(segment_id_t segment) {
+    assert(segments.size() > segment);
+    if (init_complete) {
+      assert(segments[segment].is_open());
+    } else {
+      assert(segments[segment].is_empty());
```

Comment:
I got the same assertion failure here which happens before my tree test is running, please also refer to make check failure with `unittest_extmap_manager`. Anything missing during seastore initiation?

Reply:
For initialization, you'll want to imitate unittest_transaction_manager.  I can grab your branch and have a look if you want -- is it pushed to your PR?

Reply:
Sorry, it's broken in this PR.  I'll have a look.